File: server/dukan/domain/identity.py
"""Identity & Access domain. Mirrors packages/dukan_core/lib/domain/identity.dart
and docs/domain/identity-access.md — the SAME rules and test table in both
languages. Pure: imports only the shared error contract.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from dukan.shared.errors import ConflictError, ValidationError


class Permission(StrEnum):
    SALE_CREATE = "sale.create"
    PRICE_CHANGE = "price.change"
    STOCK_ADJUST = "stock.adjust"
    PRODUCT_MANAGE = "product.manage"
    USER_MANAGE = "user.manage"
    REPORT_VIEW = "report.view"
    BRANCH_MANAGE = "branch.manage"
    DEBT_WRITE_OFF = "debt.write_off"
    AUDIT_VIEW = "audit.view"
    SALE_VOID = "sale.void"
    SALE_DISCOUNT = "sale.discount"
    CUSTOMER_CREDIT = "customer.credit"
    PURCHASE_COST = "purchase.cost"
    SETTINGS_MANAGE = "settings.manage"  # a device's own settings (idle lock)


# Built-in role name -> permission set. Single source for PermissionPolicy.
BUILTIN_ROLE_PERMISSIONS: dict[str, frozenset[Permission]] = {
    "owner": frozenset(Permission),
    "manager": frozenset({
        Permission.SALE_CREATE, Permission.PRICE_CHANGE, Permission.STOCK_ADJUST,
        Permission.PRODUCT_MANAGE, Permission.REPORT_VIEW, Permission.DEBT_WRITE_OFF,
        Permission.SALE_VOID, Permission.SALE_DISCOUNT, Permission.CUSTOMER_CREDIT,
        Permission.PURCHASE_COST, Permission.SETTINGS_MANAGE,
    }),
    "cashier": frozenset({Permission.SALE_CREATE}),
    "stock_keeper": frozenset({Permission.STOCK_ADJUST}),
    "accountant": frozenset({Permission.REPORT_VIEW, Permission.DEBT_WRITE_OFF}),
}


class UserStatus(StrEnum):
    ACTIVE = "active"
    DISABLED = "disabled"


@dataclass(frozen=True, slots=True)
class BranchAssignment:
    branch_id: str
    role_name: str


@dataclass(frozen=True, slots=True)
class User:
    id: str
    username: str
    display_name: str
    status: UserStatus
    assignments: tuple[BranchAssignment, ...] = ()
    default_branch_id: str | None = None
    version: int = 1

    @property
    def is_active(self) -> bool:
        return self.status is UserStatus.ACTIVE

    @property
    def is_owner(self) -> bool:
        return any(a.role_name == "owner" for a in self.assignments)


@dataclass(frozen=True, slots=True)
class PermissionPolicy:
    """The only place access is decided. Pure."""

    def permissions_for(self, user: User, branch_id: str) -> frozenset[Permission]:
        perms: set[Permission] = set()
        for a in user.assignments:
            if a.branch_id == branch_id:
                perms |= BUILTIN_ROLE_PERMISSIONS.get(a.role_name, frozenset())
        return frozenset(perms)

    def can(self, user: User, permission: Permission, branch_id: str) -> bool:
        if not user.is_active:
            return False
        return permission in self.permissions_for(user, branch_id)


def assert_branch_keeps_owner(*, branch_id: str, owners_after: int) -> None:
    """Every branch keeps at least one active owner, so someone can always
    administer it. `owners_after` counts the branch's active owners once the
    change applies. Raises ConflictError USER_LAST_OWNER."""
    if owners_after < 1:
        raise ConflictError("USER_LAST_OWNER", branch_id=branch_id)


def assert_shop_keeps_owner(*, owners_after: int) -> None:
    """Some active user keeps owning every branch, or nobody could open a branch
    or read the shop-wide audit trail. `owners_after` counts such users once the
    change applies. Raises ConflictError USER_LAST_OWNER."""
    if owners_after < 1:
        raise ConflictError("USER_LAST_OWNER", scope="shop")


def assert_keeps_an_assignment(*, user_id: str, remaining: int) -> None:
    """A user keeps at least one branch assignment. Raises ConflictError
    USER_LAST_ASSIGNMENT."""
    if remaining < 1:
        raise ConflictError("USER_LAST_ASSIGNMENT", user_id=user_id)


def assert_role_known(*, role_name: str) -> None:
    """Only built-in roles can be assigned until custom roles exist. Raises
    ValidationError ROLE_UNKNOWN."""
    if role_name not in BUILTIN_ROLE_PERMISSIONS:
        raise ValidationError("ROLE_UNKNOWN", role=role_name)


def assert_username_available(*, username: str, taken: bool) -> None:
    if taken:
        raise ConflictError("USER_DUPLICATE_USERNAME", username=username)


def assert_role_mutable(*, role_name: str) -> None:
    if role_name in BUILTIN_ROLE_PERMISSIONS:
        raise ConflictError("ROLE_BUILTIN_IMMUTABLE", role=role_name)


def assert_password_strong(*, password: str, min_length: int = 8) -> None:
    """A password must meet the minimum strength before it is hashed.
    Raises ValidationError WEAK_PASSWORD."""
    if len(password) < min_length:
        raise ValidationError("WEAK_PASSWORD", min_length=min_length)

File: packages/dukan_core/lib/domain/identity.dart
import '../shared/errors.dart';

/// Identity & Access domain. Mirrors docs/domain/identity-access.md and
/// server/dukan/domain/identity.py (the SAME rules run in both languages).
/// Pure: imports only the shared error contract.

/// Permissions — dotted codes. The set grows per phase.
enum Permission {
  saleCreate('sale.create'),
  priceChange('price.change'),
  stockAdjust('stock.adjust'),
  productManage('product.manage'),
  userManage('user.manage'),
  reportView('report.view'),
  branchManage('branch.manage'),
  debtWriteOff('debt.write_off'),
  auditView('audit.view'),
  saleVoid('sale.void'),
  saleDiscount('sale.discount'),
  customerCredit('customer.credit'),
  purchaseCost('purchase.cost'),

  /// This device's own settings, such as how soon an idle app locks.
  settingsManage('settings.manage');

  const Permission(this.code);
  final String code;
}

/// Built-in roles.
enum BuiltinRole {
  owner,
  manager,
  cashier,
  stockKeeper,
  accountant;

  /// Resolve from the stored role name (`owner`, `stock_keeper`, …).
  static BuiltinRole? fromName(String name) {
    switch (name) {
      case 'owner':
        return BuiltinRole.owner;
      case 'manager':
        return BuiltinRole.manager;
      case 'cashier':
        return BuiltinRole.cashier;
      case 'stock_keeper':
        return BuiltinRole.stockKeeper;
      case 'accountant':
        return BuiltinRole.accountant;
    }
    return null;
  }
}

/// The built-in role → permission mapping. Single source for [PermissionPolicy].
/// Owner is every permission (so new permissions are covered automatically).
final Map<BuiltinRole, Set<Permission>> kBuiltinRolePermissions = {
  BuiltinRole.owner: Permission.values.toSet(),
  BuiltinRole.manager: {
    Permission.saleCreate,
    Permission.priceChange,
    Permission.stockAdjust,
    Permission.productManage,
    Permission.reportView,
    Permission.debtWriteOff,
    Permission.saleVoid,
    Permission.saleDiscount,
    Permission.customerCredit,
    Permission.purchaseCost,
    Permission.settingsManage,
  },
  BuiltinRole.cashier: {Permission.saleCreate},
  BuiltinRole.stockKeeper: {Permission.stockAdjust},
  BuiltinRole.accountant: {Permission.reportView, Permission.debtWriteOff},
};

enum UserStatus { active, disabled }

/// A user's role assignment within a single branch.
final class BranchAssignment {
  const BranchAssignment({required this.branchId, required this.roleName});
  final String branchId;
  final String roleName;
}

final class User {
  const User({
    required this.id,
    required this.username,
    required this.displayName,
    required this.status,
    required this.assignments,
    this.defaultBranchId,
    this.version = 0,
  });

  final String id;
  final String username;
  final String displayName;
  final UserStatus status;
  final List<BranchAssignment> assignments;
  final String? defaultBranchId;
  final int version;

  bool get isActive => status == UserStatus.active;
  bool get isOwner => assignments.any((a) => a.roleName == 'owner');
}

/// The only place access is decided. Pure.
final class PermissionPolicy {
  const PermissionPolicy();

  /// Effective permissions a user has in [branchId] (union across that
  /// branch's role assignments).
  Set<Permission> permissionsFor(User user, String branchId) {
    final perms = <Permission>{};
    for (final a in user.assignments.where((a) => a.branchId == branchId)) {
      final role = BuiltinRole.fromName(a.roleName);
      if (role != null) perms.addAll(kBuiltinRolePermissions[role] ?? const {});
    }
    return perms;
  }

  bool can(User user, Permission permission, String branchId) {
    if (!user.isActive) return false;
    return permissionsFor(user, branchId).contains(permission);
  }
}

/// Domain invariant: every branch keeps at least one active owner, so someone
/// can always administer it. [ownersAfter] counts the branch's active owners
/// once the change applies. Raises [ConflictError] `USER_LAST_OWNER`.
void assertBranchKeepsOwner({required String branchId, required int ownersAfter}) {
  if (ownersAfter < 1) {
    throw ConflictError('USER_LAST_OWNER', {'branch_id': branchId});
  }
}

/// Domain invariant: some active user keeps owning every branch, or nobody
/// could open a branch or read the shop-wide audit trail. [ownersAfter] counts
/// such users once the change applies. Raises [ConflictError] `USER_LAST_OWNER`.
void assertShopKeepsOwner({required int ownersAfter}) {
  if (ownersAfter < 1) {
    throw ConflictError('USER_LAST_OWNER', const {'scope': 'shop'});
  }
}

/// Domain invariant: a user keeps at least one branch assignment.
/// Raises [ConflictError] `USER_LAST_ASSIGNMENT`.
void assertKeepsAnAssignment({required String userId, required int remaining}) {
  if (remaining < 1) {
    throw ConflictError('USER_LAST_ASSIGNMENT', {'user_id': userId});
  }
}

/// Only built-in roles can be assigned until custom roles exist.
/// Raises [ValidationError] `ROLE_UNKNOWN`.
void assertRoleKnown({required String roleName}) {
  if (BuiltinRole.fromName(roleName) == null) {
    throw ValidationError('ROLE_UNKNOWN', {'role': roleName});
  }
}

/// Domain invariant: usernames are unique among active users.
/// Raises [ConflictError] `USER_DUPLICATE_USERNAME`.
void assertUsernameAvailable({
  required String username,
  required bool taken,
}) {
  if (taken) {
    throw ConflictError('USER_DUPLICATE_USERNAME', {'username': username});
  }
}

/// Domain invariant: a built-in role's core permissions are immutable.
/// Raises [ConflictError] `ROLE_BUILTIN_IMMUTABLE`.
void assertRoleMutable({required String roleName}) {
  if (BuiltinRole.fromName(roleName) != null) {
    throw ConflictError('ROLE_BUILTIN_IMMUTABLE', {'role': roleName});
  }
}

/// Security invariant: a password must meet the minimum strength before it is
/// hashed. Raises [ValidationError] `WEAK_PASSWORD`.
void assertPasswordStrong({required String password, int minLength = 8}) {
  if (password.length < minLength) {
    throw ValidationError('WEAK_PASSWORD', {'min_length': minLength});
  }
}
